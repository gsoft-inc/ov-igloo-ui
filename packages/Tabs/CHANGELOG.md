# @igloo-ui/tabs

## 0.1.0

### Minor Changes

- fc12e4e: Initial release of the tabs component
