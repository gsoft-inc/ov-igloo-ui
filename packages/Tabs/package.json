{
  "name": "@igloo-ui/tabs",
  "version": "1.2.1",
  "main": "dist/Tabs.js",
  "module": "dist/Tabs.js",
  "types": "dist/Tabs.d.ts",
  "styles": "dist/tabs.css",
  "publishConfig": {
    "access": "public"
  },
  "repository": "git+https://github.com/gsoft-inc/ov-igloo-ui.git",
  "author": "Officevibe Inc.",
  "contributors": [
    "Vicky Comeau vicky.comeau@gsoft.com"
  ],
  "license": "Apache-2.0",
  "peerDependencies": {
    "react": ">= 16.8.6"
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "rollup -c rollup.config.js"
  },
  "dependencies": {
    "@igloo-ui/tokens": "^2.0.0",
    "@hopper-ui/tokens": "^3.1.3",
    "classnames": "^2.3.2",
    "@igloo-ui/icons": "^1.12.2"
  },
  "devDependencies": {
    "react-router-dom": "^6.3.0"
  },
  "browserslist": [
    "> 1%",
    "last 2 versions",
    "IE >= 11"
  ]
}
