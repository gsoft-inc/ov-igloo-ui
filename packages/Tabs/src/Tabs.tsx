import * as React from "react";

import cx from "classnames";

import CrownSolid from "@igloo-ui/icons/dist/CrownSolid";
import TextBulletSolid from "@igloo-ui/icons/dist/TextBulletSolid";
import { BulletIcon, UpsellIcon } from "@hopper-ui/icons-react16";

import "./tabs.scss";

export interface TabInterface {
    /** The contents of the tabs list */
    children?: React.ReactNode;
    /** Add an id attribute to the tab */
    id?: string;
    /** Add text for the tab */
    label: React.ReactElement | string;
    /** Add a notification indicator to the tab */
    notification?: boolean;
    /** Add a premium indicator to the tab */
    premium?: boolean;
}

export interface TabsProps extends React.ComponentProps<"div"> {
    /** The function that's called when a tab is clicked */
    onSelectTab?: (index: number) => void;
    /** The index number of the tab that should be selected on load */
    selected?: number;
    /** Add a data-test tag for automated tests */
    dataTest?: string;
    /** Add a specific class to the tabs component */
    className?: string;
    /** Add an inline style to the tabs */
    isInline?: boolean;
    /** An array of objects that specify the props of the tab buttons */
    tabs: Array<TabInterface>;
}

const getBrand = (): string => {
    return document.documentElement.getAttribute("data-brand") ?? "igloo";
};

const Tabs: React.FunctionComponent<TabsProps> = ({
    onSelectTab,
    selected = 0,
    dataTest,
    className,
    isInline = true,
    tabs,
    ...rest
}) => {
    const classes = cx("ids-tabs", className, {
        "ids-tabs--inline": isInline,
        "ids-tabs--heading": !isInline
    });
    
    const isWorkleap = getBrand() === "workleap";

    const handleOnClick = (index: number): void => {
        if (onSelectTab) {
            onSelectTab(index);
        }
    };

    const renderTabItem = (tab: TabInterface, index: number): React.ReactNode => {
        const isTypeString = typeof tab.label === "string";
        const notificationClass = "ids-tab__icon ids-tab__bullet";
        const premiumClass = "ids-tab__icon ids-tab__crown";
        const tabTextClass = "ids-tab__text";
        let tabContents = (
            <>
                {tab.notification && (
                    isWorkleap ? (
                        <BulletIcon
                            size="sm"
                            className={notificationClass}
                        />
                    ) : (
                        <TextBulletSolid
                            size="small"
                            className={notificationClass}
                        />
                    )
                )}
                {tab.premium && (
                    isWorkleap ? (
                        <UpsellIcon
                            size="sm"
                            className={premiumClass}
                        />
                    ) : (
                        <CrownSolid
                            size="small"
                            className={premiumClass}
                        />
                    )
                )}
            </>
        );

        if (isTypeString) {
            tabContents = (
                <button className="ids-tab__btn" onClick={() => handleOnClick(index)}>
                    <span className={tabTextClass}>{tab.label}</span>
                    {tabContents}
                </button>
            );
        } else if (React.isValidElement(tab.label)) {
            const textWrapper = <span className={tabTextClass}>
                {(tab.label as React.ReactElement).props.children}
            </span>;
            tabContents = <span className="ids-tab__nav">
                {React.cloneElement(tab.label,
                                    {},
                                    textWrapper, tabContents)
                }
            </span>;
        }

        return (
            <li
                key={index.toString()}
                className={cx("ids-tab", {
                    "ids-tab--active": index === selected
                })}
                id={tab.id}
            >
                {tabContents}
            </li>
        );
    };

    return (
        <div className={classes} data-test={dataTest} {...rest}>
            <ul className="ids-tabs__list">
                {tabs.map((tab, index) => renderTabItem(tab, index))}
            </ul>

            {tabs[selected].children && tabs[selected].children}
        </div>
    );
};

export default Tabs;
