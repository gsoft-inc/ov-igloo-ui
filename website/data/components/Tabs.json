[
  [
    {
      "tags": {},
      "filePath": "/Users/franck.gaudin/Devel/Igloo/components/packages/Tabs/src/Tabs.tsx",
      "description": "",
      "displayName": "Tabs",
      "methods": [],
      "props": {
        "onSelectTab": {
          "defaultValue": null,
          "description": "The function that's called when a tab is clicked",
          "name": "onSelectTab",
          "parent": {
            "fileName": "components/packages/Tabs/src/Tabs.tsx",
            "name": "TabsProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Tabs/src/Tabs.tsx",
              "name": "TabsProps"
            }
          ],
          "required": false,
          "type": { "name": "(index: number) => void" }
        },
        "selected": {
          "defaultValue": { "value": 0 },
          "description": "The index number of the tab that should be selected on load",
          "name": "selected",
          "parent": {
            "fileName": "components/packages/Tabs/src/Tabs.tsx",
            "name": "TabsProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Tabs/src/Tabs.tsx",
              "name": "TabsProps"
            }
          ],
          "required": false,
          "type": { "name": "number" }
        },
        "dataTest": {
          "defaultValue": null,
          "description": "Add a data-test tag for automated tests",
          "name": "dataTest",
          "parent": {
            "fileName": "components/packages/Tabs/src/Tabs.tsx",
            "name": "TabsProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Tabs/src/Tabs.tsx",
              "name": "TabsProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "className": {
          "defaultValue": null,
          "description": "Add a specific class to the tabs component",
          "name": "className",
          "parent": {
            "fileName": "components/packages/Tabs/src/Tabs.tsx",
            "name": "TabsProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Tabs/src/Tabs.tsx",
              "name": "TabsProps"
            }
          ],
          "required": false,
          "type": { "name": "string" }
        },
        "isInline": {
          "defaultValue": { "value": true },
          "description": "Add an inline style to the tabs",
          "name": "isInline",
          "parent": {
            "fileName": "components/packages/Tabs/src/Tabs.tsx",
            "name": "TabsProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Tabs/src/Tabs.tsx",
              "name": "TabsProps"
            }
          ],
          "required": false,
          "type": { "name": "boolean" }
        },
        "tabs": {
          "defaultValue": null,
          "description": "An array of objects that specify the props of the tab buttons",
          "name": "tabs",
          "parent": {
            "fileName": "components/packages/Tabs/src/Tabs.tsx",
            "name": "TabsProps"
          },
          "declarations": [
            {
              "fileName": "components/packages/Tabs/src/Tabs.tsx",
              "name": "TabsProps"
            }
          ],
          "required": true,
          "type": { "name": "TabInterface[]" }
        }
      }
    }
  ]
]
