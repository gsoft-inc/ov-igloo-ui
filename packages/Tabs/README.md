# Tabs

Use to alternate among related views within the same context.

<Example />

<ReferenceLinks />

## Installation

To install `@igloo-ui/tabs` in your project, you will need to run the following command using [npm](https://www.npmjs.com/):

```bash
npm install @igloo-ui/tabs
```

If you prefer [Yarn](https://classic.yarnpkg.com/en/), use the following command instead:

```bash
yarn add @igloo-ui/tabs
```

## Usage

Then to use the component in your code just import it!

```jsx
import Tabs, { TabPane } from '@igloo-ui/tabs';

<Tabs
  defaultIndex={1}
  tabs={[
    {
      label: 'Premium Tab',
      premium: true,
      children: 'Content of premium tab',
    },
    {
      label: 'Notification Tab',
      notification: true,
      children: 'Content of notification tab',
    },
    {
      label: 'Premium Notification Tab',
      notification: true,
      premium: true,
      children: 'Content of premium notification tab',
    },
  ]}
/>;
```
