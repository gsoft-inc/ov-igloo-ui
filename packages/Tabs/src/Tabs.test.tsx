/**
 * @jest-environment jsdom
 */
import React from 'react';
import { render, RenderResult, screen } from '@testing-library/react';

import Tabs, { TabsProps } from './Tabs';

const setup = (props: TabsProps) => {
  return render(<Tabs {...props} />);
};

const dataTestId = 'ids-tabs-id';

describe('Tabs', () => {
  const defaultComponentSetUp: TabsProps = {
    dataTest: dataTestId,
    tabs: [
      {
        label: 'tab 1',
        children: 'Tab 1 content',
      },
      {
        label: 'tab 2',
        children: 'Tab 2 content',
      },
    ],
  };
  const component = () => setup(defaultComponentSetUp);

  const expectToBeOfStyle = (component: RenderResult, inline: boolean) => {
    const tabs = component.container.firstChild;

    const expectedClass = 'ids-tabs--' + (inline ? 'inline' : 'heading');
    expect(tabs).toHaveClass(expectedClass);
  };

  test('It should render without errors', () => {
    component();
    const tabs = screen.getByTestId(dataTestId);

    expect(tabs).toBeInTheDocument();
  });

  test('It should render a snapshot', () => {
    const { asFragment } = component();
    expect(asFragment()).toMatchSnapshot();
  });

  test('It should render an inline style', () => {
    const inline = true;
    const tabs1 = component();
    const tabs2 = setup({ ...defaultComponentSetUp, isInline: inline });

    // Check without setting isInline and setting it to true.
    expectToBeOfStyle(tabs1, inline);
    expectToBeOfStyle(tabs2, inline);
  });

  test('It should render a heading style', () => {
    const inline = false;
    const tabs = setup({ ...defaultComponentSetUp, isInline: inline });

    expectToBeOfStyle(tabs, inline);
  });

  test('It should render a tab with a premium icon', () => {
    let tabsArray = [...defaultComponentSetUp.tabs];
    tabsArray[0].premium = true;
    const { container } = setup({
      ...defaultComponentSetUp,
      tabs: tabsArray,
    });
    const premiumIcon = container.querySelector('.ids-tab__crown');

    expect(premiumIcon).toBeInTheDocument();
  });

  test('It should render a tab with a notification icon', () => {
    let tabsArray = [...defaultComponentSetUp.tabs];
    tabsArray[0].notification = true;
    const { container } = setup({
      ...defaultComponentSetUp,
      tabs: tabsArray,
    });
    const notificationIcon = container.querySelector('.ids-tab__bullet');

    expect(notificationIcon).toBeInTheDocument();
  });

  test('It should render tabs with a default index of 1', () => {
    const { container } = setup({ ...defaultComponentSetUp, selected: 1 });

    const secondTab = container.querySelectorAll('.ids-tab')[1];

    expect(secondTab).toHaveClass('ids-tab--active');
  });
});
