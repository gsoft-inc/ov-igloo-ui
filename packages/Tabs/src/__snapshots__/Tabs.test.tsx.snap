// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs It should render a snapshot 1`] = `
<DocumentFragment>
  <div
    class="ids-tabs ids-tabs--inline"
    data-test="ids-tabs-id"
  >
    <ul
      class="ids-tabs__list"
    >
      <li
        class="ids-tab ids-tab--active"
      >
        <button
          class="ids-tab__btn"
        >
          tab 1
        </button>
      </li>
      <li
        class="ids-tab"
      >
        <button
          class="ids-tab__btn"
        >
          tab 2
        </button>
      </li>
    </ul>
    Tab 1 content
  </div>
</DocumentFragment>
`;
